File: py/aoc2023/day15.py
"""
Day 15: Lens Library
"""

from functools import reduce

SAMPLE_INPUT = """rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7"""


def _hash(string):
    return reduce(lambda acc, char: 17 * (acc + ord(char)) % 256, string, 0)


def part1(data: str):
    """
    >>> part1(SAMPLE_INPUT)
    1320
    """
    return sum(map(_hash, data.strip().split(",")))


def part2(data):
    """
    >>> part2(SAMPLE_INPUT)
    145
    """
    buckets = [{} for _ in range(256)]
    for step in data.strip().split(","):
        if step.endswith("-"):
            key = step[:-1]
            buckets[_hash(key)].pop(key, None)
        elif "=" in step:
            key = step[: step.index("=")]
            buckets[_hash(key)][key] = int(step[step.index("=") + 1 :])
    return sum(
        (i + 1) * sum((j + 1) * n for j, n in enumerate(bucket.values()))
        for i, bucket in enumerate(buckets)
    )


parts = (part1, part2)

File: hs/src/Day15.hs
{-|
Module:         Day15
Description:    <https://adventofcode.com/2023/day/15 Day 15: Lens Library>
-}
{-# LANGUAGE OverloadedStrings, TransformListComp #-}
module Day15 (part1, part2) where

import Data.Bits ((.&.))
import Data.Char (isAlphaNum, isSpace, ord)
import Data.Function (on)
import Data.List (foldl', groupBy, sortOn)
import qualified Data.Map as Map (delete, empty, insertWith, toList)
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import qualified Data.Text as T (foldl', null, span, split, stripPrefix, stripSuffix)
import qualified Data.Text.Read as T (decimal)
import GHC.Exts (the)

hash :: Text -> Int
hash = (.&. 255) . T.foldl' hash' 0 where hash' k c = 17 * (k + ord c)

isSep :: Char -> Bool
isSep ',' = True
isSep c = isSpace c

part1, part2 :: Text -> Int
part1 = sum . map hash . T.split isSep
part2 = power . foldl' go Map.empty . zip [0..] . filter (not . T.null) . T.split isSep
  where
    go lens (c, ins)
      | "-" <- rest = Map.delete key lens
      | Just (Right (n, "")) <- T.decimal <$> T.stripPrefix "=" rest
      = Map.insertWith (\(n, _) (_, c) -> (n, c)) key (n, c) lens
      where (key, rest) = T.span isAlphaNum ins
    power lens = sum
      [ (the (hash <$> key) + 1) * sum (zipWith (*) [1..] n)
      | (key, (n, c)) <- Map.toList lens
      , then sortOn by (hash key, c)
      , then group by hash key using groupBy . on (==)
      ]
